### PR TITLE
Add support for alternative ID types

### PR DESCRIPTION
**Pull request by stwiname**
# Description
Allows specifying a different ID type other than text. It can be 'BigInt', 'Int', 'Float', 'ID' or 'String'. This is useful when you want to order by ID but not in a lexicographical order. 

In order to avoid growing this PR and the complexity of types, codegen has been updated to generate some `Compat` types to work with internal types that expect the `id` to be a string.

Fixes https://github.com/subquery/subql/issues/2562
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update https://github.com/subquery/documentation/pull/577

## Checklist

- [x] I have tested locally
- [x] I have performed a self review of my changes
- [x] Updated any relevant documentation https://github.com/subquery/documentation/pull/577
- [x] Linked to any relevant issues
- [x] I have added tests relevant to my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My code is up to date with the base branch
- [x] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/cli/src/template/model.ts.ejs [hunk 1]
```diff
@@ -56,14 +62,14 @@ export class <%= props.className %> implements Entity {
     static async getBy<%=helper.upperFirst(field.name) %>(<%=field.name %>: <%=field.type %>): Promise<<%=props.className %> | undefined> {
         const record = await store.getOneByField('<%=props.entityName %>', '<%=field.name %>', <%=field.name %>);
         if (record) {
-            return this.create(record as <%= props.className %>Props);
+            return this.create(record as unknown <%= props.className %>Props);
```
Comment (yoozo) [inline packages/cli/src/template/model.ts.ejs:65]: Missing one as

File: packages/cli/src/template/model.ts.ejs [hunk 2]
```diff
@@ -31,21 +37,21 @@ export class <%= props.className %> implements Entity {
     }
 
     async save(): Promise<void> {
-        let id = this.id;
+        const id = this.id;
         assert(id !== null, "Cannot save <%=props.className %> entity without an ID");
-        await store.set('<%=props.entityName %>', id.toString(), this);
+        await store.set('<%=props.entityName %>', id.toString(), this as Compat<%=props.className %>Props);
```
Comment (yoozo) [inline packages/cli/src/template/model.ts.ejs:42]: The type issue of the id field still needs to be addressed here.

` {...this, id: id.toString() } as CompatTransferProps`

**Review by yoozo — COMMENTED** [review]
There will be a type error when using the `store.bulkCreate` method because it requires the `entity.id` type to be a string.

File: packages/utils/src/graphql/graphql.spec.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
 // SPDX-License-Identifier: GPL-3.0
 
+import {buildSchema} from 'graphql';
```
Comment (yoozo) [inline packages/utils/src/graphql/graphql.spec.ts:4]: can remove it

File: packages/cli/src/template/model.ts.ejs (hunk 2 above)
Comment (stwiname) [reply packages/cli/src/template/model.ts.ejs:42]: Functionally this is fine, its just the types we're getting around. Its better to not copy the data if we can avoid it

**Comment by stwiname** [comment]
> There will be a type error when using the `store.bulkCreate` method because it requires the `entity.id` type to be a string.

I think that is fine. People don't generally have much use for `bulkCreate` and if the cache is enabled there is no performance benefit. I want to avoid going down the rabbit hole of trying to update all the types.

**Review by yoozo — APPROVED** [review]

**Pull request merged**